### PR TITLE
[enhancement](Nereids) make HyperGraph can be built from plan

### PR DESCRIPTION
**Pull request by keanji-x**
## Proposed changes

make HyperGraph can be built from plan

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by XieJiann** [comment]
run buildall

**Comment by XieJiann** [comment]
run buildall

**Comment by XieJiann** [comment]
run buildall

**Comment by XieJiann** [comment]
run buildall

**Comment by XieJiann** [comment]
run buildall

**Comment by XieJiann** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/nereids/jobs/joinorder/hypergraph/HyperGraph.java [hunk 1]
```diff
@@ -304,6 +304,76 @@ private long calNodeMap(Set<Slot> slots) {
         return bitmap;
     }
 
+    public static HyperGraph from(Plan plan) {
+        Preconditions.checkArgument(plan.getGroupExpression().isPresent(),
```
Comment (seawinde) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/jobs/joinorder/hypergraph/HyperGraph.java:308]: Should support plan without group. May be inside the methed from(Plan plan), call the memo.init(), and then build graph

**Comment by XieJiann** [comment]
run buildall

**Comment by XieJiann** [comment]
run buildall

**Comment by XieJiann** [comment]
run buildall

**Review by seawinde — APPROVED** [review]

File: fe/fe-core/src/main/java/org/apache/doris/nereids/jobs/joinorder/hypergraph/HyperGraph.java [hunk 2]
```diff
@@ -304,6 +320,108 @@ private long calNodeMap(Set<Slot> slots) {
...
+        return Pair.of(new BitSet(), LongBitmap.newBitmap(idx));
+    }
+
+    // Build Graph for matching mv
+    private Pair<BitSet, Long> buildStructInfo(Plan plan) {
+        if (plan instanceof GroupPlan) {
+            Group group = ((GroupPlan) plan).getGroup();
+            if (group.getHyperGraph() == null) {
+                buildStructInfo(group.getLogicalExpressions().get(0).getPlan());
+            } else {
```
Comment (XieJiann) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/jobs/joinorder/hypergraph/HyperGraph.java:371]: Add with GroupPlan children

File: fe/fe-core/src/main/java/org/apache/doris/nereids/jobs/joinorder/hypergraph/HyperGraph.java [hunk 3]
```diff
@@ -304,6 +320,108 @@ private long calNodeMap(Set<Slot> slots) {
         return bitmap;
     }
 
+    public static HyperGraph toStructInfo(Plan plan) {
+        Preconditions.checkArgument(plan.getGroupExpression().isPresent(),
```
Comment (XieJiann) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/jobs/joinorder/hypergraph/HyperGraph.java:324]: multiple HyperGraph

**Review by jackwener — APPROVED** [review]

**Pull request merged**